plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.jvm.abi
+
+import kotlinx.metadata.*
+import kotlinx.metadata.jvm.JvmClassExtensionVisitor
+import kotlinx.metadata.jvm.KotlinClassHeader
+import kotlinx.metadata.jvm.KotlinClassMetadata
+import org.jetbrains.kotlin.load.java.JvmAnnotationNames
+import org.jetbrains.org.objectweb.asm.AnnotationVisitor
+import org.jetbrains.org.objectweb.asm.Opcodes
+
+/**
+ * Wrap the visitor for a Kotlin Metadata annotation to strip out private and local
+ * functions, properties, and type aliases as well as local delegated properties.
+ */
+fun abiMetadataProcessor(annotationVisitor: AnnotationVisitor): AnnotationVisitor =
+    kotlinClassHeaderVisitor { header ->
+        val newHeader = when (val metadata = KotlinClassMetadata.read(header)) {
+            is KotlinClassMetadata.Class -> {
+                val writer = KotlinClassMetadata.Class.Writer()
+                metadata.accept(AbiKmClassVisitor(writer))
+                writer.write(header.metadataVersion, header.extraInt).header
+            }
+            is KotlinClassMetadata.FileFacade -> {
+                val writer = KotlinClassMetadata.FileFacade.Writer()
+                metadata.accept(AbiKmPackageVisitor(writer))
+                writer.write(header.metadataVersion, header.extraInt).header
+            }
+            is KotlinClassMetadata.MultiFileClassPart -> {
+                val writer = KotlinClassMetadata.MultiFileClassPart.Writer()
+                metadata.accept(AbiKmPackageVisitor(writer))
+                writer.write(metadata.facadeClassName, header.metadataVersion, header.extraInt).header
+            }
+            else -> header
+        }
+
+        // Write out the stripped annotation
+        annotationVisitor.visitKotlinMetadata(newHeader)
+    }
+
+/**
+ * Parse a KotlinClassHeader from an existing Kotlin Metadata annotation visitor.
+ */
+private fun kotlinClassHeaderVisitor(body: (KotlinClassHeader) -> Unit): AnnotationVisitor =
+    object : AnnotationVisitor(Opcodes.API_VERSION) {
+        var kind: Int = 1
+        var metadataVersion: IntArray = intArrayOf()
+        var bytecodeVersion: IntArray = intArrayOf()
+        var data1: MutableList<String> = mutableListOf()
+        var data2: MutableList<String> = mutableListOf()
+        var extraString: String = ""
+        var packageName: String = ""
+        var extraInt: Int = 0
+
+        override fun visit(name: String, value: Any?) {
+            when (name) {
+                "k" -> kind = value as Int
+                "extraInt" -> extraInt = value as Int
```

Comment:
I think it should be `"xi"`

Reply:
Right... I switched to using the constants from `JvmAnnotationNames`.